feat(lib): use inject to pass container reference to selectableItem

bvkimball opened the pull request:
Needed to use `[dtsSelectItem]` in grandchildren components which can't be found `@ContentChildren`

Now using dependency injection i pass the reference of the `SelectContainerComponent` to any `SelectItemDirective` that is a child of that container regardless of how deep in the tree the component is.

This is similar to how material does this for drag and drop components for the CDK. see [here](https://github.com/angular/components/tree/master/src/cdk/drag-drop/directives)

Just pushing this out to see if there is interest in adopting this change, currently the demo app work as is, but I can update the demo app with an example of nested components using the directive and do a little more cleanup.

- [x] clean up the code
- [x] make sure all tests pass
- [x] maybe create an additional example section to show this in action
- [x] update the readme accordingly
- [x] write tests for this change (if needed)


> ps. awesome library

Commit:
feat(lib): use inject to pass container reference to selectableItem

d3lm commented:
Hey @bvkimball, thanks for your PR. This is actually something I was fiddling with as well at some point but it never made it into master. I like this refactoring and agree with what you outlined above. I am definitely down to adopting this PR if you:

- clean up the code
- make sure all tests pass
- maybe create an additional example section to show this in action
- update the readme accordingly
- write tests for this change (if needed)

How does that sound?

Thanks for your contribution 👍 and I am glad to hear you like this library and it's useful to you.

bvkimball commented:
Awesome! I will try to get this cleaned up over the weekend.

d3lm commented:
Love it. Thanks a bunch!

Commit:
docs: added nested directive demo with drag and drop
refactor: removed dead code from change and fixed circular dependency caused by DI

bvkimball commented:
Not perfect yet but I think tests are working again.  Issue caused because the Parent-Child Injection causes circular dependency.

The demo is a bit contrived but it is similar to the usecase I am using the library for.  I intend to use it with the CDK DragAndDropModule, so that what the demo does.  Here is a gif of the demo.

![demo](http://g.recordit.co/3ZI1kBWDvf.gif)

When I re-order the **tasks** then try to use the select box it selects the task based on the previous position, i tried running selectionContainer.update() but that didn't fix the issue.  I am trying to make sure this isn't something i caused with the change or the demo code itself.

Another issue I am looking at is that click to select doesn't seem to work, i think that might be a conflict with Material or the Drag&Drop Module too.

Commit:
style: fixed lint error for member ordering

d3lm commented:
Yep that demo is looking slick! Thanks for this. Will be a great addition. Make sure to not break existing functionality. As to the "click to select", I am not sure. Could be indeed a conflict with the Drag & Drop module.

d3lm commented:
Let me know when this PR is ready for review and I ll look at it. Maybe you can also take the list of action items I posted earlier and make a small todo list, so we keep track of what's done.

Commit:
docs: added new section to README

Commit:
docs(lib): added section to readme about components capturing events
test(app): added cypress test for new example demo

bvkimball commented:
I think everything is done and working, and I add docs and some cypress test.
The issues with select are caused by the Angular CDK DragDrop swallows all mouse events.
Let me know about any changes or if you want me to rename any of the methods or tokens.

I believe there are no breaking changes unless someone was using $selectItems which is the QueryList that was removed. that was public but really should have been used for Internal use only....

Commit:
fix(lib): new Token was not export with the public api

d3lm commented:
Yep, agreed that the QueryList shouldn't have been public in the first place. If it causes a breaking change we can revert this and publish this as a new major version.

I ll look at your changes today and review everything.

d3lm commented on README.md:
```diff
@@ -554,6 +554,26 @@ Here we listen for `scroll` on the `div` and call `container.update()` in case t
 
+### Does the `selectItem` directive need to be a direct child of the `dts-select-container`?
+
+As of version `4.1.0`, an injetion token is used to pass the SelectContainerComponent parent to the directive. You can use this in any component nested within the `dts-select-container`.
```
Wrap "SelectContainerComponent" in an inline code block.

d3lm commented on README.md:
````diff
@@ -554,6 +554,26 @@ Here we listen for `scroll` on the `div` and call `container.update()` in case t
+
+As of version `4.1.0`, an injetion token is used to pass the SelectContainerComponent parent to the directive. You can use this in any component nested within the `dts-select-container`.
+
+```javascript
````
ts

d3lm commented on README.md:
````diff
@@ -554,6 +554,26 @@ Here we listen for `scroll` on the `div` and call `container.update()` in case t
+    public container: SelectContainerComponent) {}
+```
+
+You can see an example of the this in the [drag and drop example](https://github.com/d3lm/ngx-drag-to-select/blob/master/src/app/dragndrop). You can see that the `[selectItem]` directive is set in `app-task` component and the `dts-select-container` is in the `dragndrop` component.
````
You can find an example of this...

d3lm replied:
Also remove the second sentence. I think it doesn't have a lot of value add.

d3lm commented on README.md:
```diff
@@ -554,6 +554,26 @@ Here we listen for `scroll` on the `div` and call `container.update()` in case t
+@Component({...})
+export class TaskListComponent {
+  constructor(
+    @Inject(DTS_SELECT_CONTAINER) @Optional()
```
Reformat constructor to
```
constructor(
  @Inject(DTS_SELECT_CONTAINER) @Optional()
  public container: SelectContainerComponent
) {}
```

d3lm commented on README.md:
```diff
@@ -554,6 +554,26 @@ Here we listen for `scroll` on the `div` and call `container.update()` in case t
+
+You can see an example of the this in the [drag and drop example](https://github.com/d3lm/ngx-drag-to-select/blob/master/src/app/dragndrop). You can see that the `[selectItem]` directive is set in `app-task` component and the `dts-select-container` is in the `dragndrop` component.
+
+### Why is my `selectItem` directive not selecting why I click on it?
```
...not selecting **when** I click on it?

d3lm commented on README.md:
```diff
@@ -554,6 +554,26 @@ Here we listen for `scroll` on the `div` and call `container.update()` in case t
+
+### Why is my `selectItem` directive not selecting why I click on it?
+
+If you are using the `selectItem` within a nested component then your mousedown/up events might being captured by another directive or component in your code. For example if you are using this libary with Angular CDK's DragDropModule, the mouse events are captured by the `cdkDrag` directive, [see here](https://github.com/angular/components/pull/19674).
```
`DragDropModule`

d3lm commented on cypress/integration/drag-and-drop.spec.ts:
```diff
@@ -0,0 +1,80 @@
+    it('should drag to new list', () => {
+      getDragAndDropExample().within(() => {
+        getTodoList()
+          // Select First 3 items in list
```
Single line comments always all lower case

d3lm replied:
Did you forget this?

d3lm commented on cypress/integration/drag-and-drop.spec.ts:
```diff
@@ -0,0 +1,80 @@
+          .getSelectItem(2)
+          .dispatch('mousemove')
+          .dispatch('mouseup')
+          // Click on second item
```
same

d3lm replied:
And here too.

d3lm commented on cypress/integration/drag-and-drop.spec.ts:
```diff
@@ -0,0 +1,80 @@
+          // Click on second item
+          .getSelectItem(1)
+          .dispatch('mousedown', { button: 0 })
+          // Drag to Select Item in other list
```
same

d3lm replied:
Same again.

d3lm commented on projects/ngx-drag-to-select/src/lib/select-container.component.ts:
```diff
@@ -281,14 +285,24 @@ export class SelectContainerComponent implements AfterViewInit, OnDestroy, After
+  register(item: SelectItemDirective) {
+    this._registry.add(item);
+    this.updateSelectableItems();
+  }
```
newline after

d3lm replied:
Still no newline? Or did the PR not update?

d3lm commented on projects/ngx-drag-to-select/src/lib/tokens.ts:
```diff
@@ -1,5 +1,7 @@
+import { ComponentType } from '@angular/cdk/portal';
```
Let's not use this, because the CDK is not a peer dependency of this library and IMO we shouldn't use anything from the CDK for the library.

bvkimball replied:
agreed this was silly of me, I just created a version of ComponentType in the models.ts, it is just a Util type anyway that wraps a interface in a constructable, hope this works.

d3lm commented on src/app/dragndrop/dragndrop.component.html:
```diff
@@ -0,0 +1,9 @@
+<dts-select-container data-cy="drag-and-drop-container" [custom]="true" [(selectedItems)]="selectedItems">
+  <div class="list-wrapper" cdkDropListGroup>
+
```
remove newline

d3lm commented on src/app/dragndrop/dragndrop.component.html:
```diff
@@ -0,0 +1,9 @@
+    <app-task-list title="Todo" [tasks]="todo" data-cy="todo-list"></app-task-list>
+    <app-task-list title="Doing" [tasks]="doing" data-cy="doing-list"></app-task-list>
+    <app-task-list title="Done" [tasks]="done" data-cy="done-list"></app-task-list>
+
```
remove newline

d3lm commented on src/app/dragndrop/task-list/task-list.component.ts:
```diff
@@ -0,0 +1,60 @@
+    this.tasks.splice(spliceIntoIndex, 0, ...this.container.selectedItems);
+
+    setTimeout(() => this.container.update());
+    // if (event.previousContainer === event.container) {
```
Remove comment

d3lm commented on src/app/dragndrop/task-list/task-list.component.ts:
```diff
@@ -0,0 +1,60 @@
+    const spliceIntoIndex = event.currentIndex;
+    this.tasks.splice(spliceIntoIndex, 0, ...this.container.selectedItems);
+
+    setTimeout(() => this.container.update());
```
Add a comment why the `setTimeout` is needed. I have a vague idea but this is useful to other people as well.

d3lm reviewed (changes requested):
It's already looking really good. I love this addition / change.

Left a few remarks though.

Also, can you change `dragndrop` to be `drag-and-drop`?

Thanks for this contribution!

d3lm commented on projects/ngx-drag-to-select/src/lib/select-container.component.ts:
```diff
@@ -259,11 +259,15 @@ export class SelectContainerComponent implements AfterViewInit, OnDestroy, After
 
   ngAfterContentInit() {
-    this._selectableItems = this.$selectableItems.toArray();
+    // removed query list, maybe remove this too
```
Can this be removed?

Commit:
refactor(app): renamed drag-and-drop
docs: updated grammar

Commit:
refactor(app): forgot to add new files

bvkimball commented:
I think i have updated everything correctly, sorry about my poor grammar skills ;)

d3lm commented on README.md:
```diff
@@ -554,6 +554,27 @@ Here we listen for `scroll` on the `div` and call `container.update()` in case t
+  constructor(
+    @Inject(DTS_SELECT_CONTAINER) @Optional()
+    public container: SelectContainerComponent
+  ) {}
```
Sorry for this nit-pick but the class is missing a closing curly brace.

d3lm reviewed (changes requested):
Left a few minor remarks, most of which were comments from my previous PR which have not been updated. Not sure if the PR was not entirely updated or if you just forgot a few comments. Not a biggy tho.

Commit:
style: updated cypress comments
style: update whitespace

bvkimball commented:
ok, sorry about those missed comments, I wish GitHub had an easy way to view them in a list.  I resolved the old outdated comments.

d3lm reviewed (approved):
This looks good now! Thanks a lot for this PR. I think it's a great addition.

d3lm commented on src/app/drag-and-drop/task/task.component.scss:
```diff
@@ -0,0 +1,8 @@
+:host {
```
Can you change this to
```
:host {
  height: min-content;
}

mat-card {
  border: 1px solid transparent;

  &.selected {
    border: 1px solid #2196f3;
  }
}
```

Because without the border, the card changes its size when selected.

d3lm commented on src/app/drag-and-drop/task-list/task-list.component.scss:
```diff
@@ -0,0 +1,69 @@
+    transform: rotate(-1deg);
+    z-index: -1;
+  }
+  mat-card:nth-child(3) {
```
Newline before.

d3lm commented on src/app/drag-and-drop/task-list/task-list.component.scss:
```diff
@@ -0,0 +1,69 @@
+    top: -10px;
+    left: -25px;
+  }
+  mat-card:nth-child(2) {
```
Newline before.

d3lm commented on src/app/drag-and-drop/task-list/task-list.component.scss:
```diff
@@ -0,0 +1,69 @@
+  display: flex;
+  flex-flow: column;
+  gap: 0.5rem;
+  mat-card {
```
Newline before.

d3lm commented on src/app/app.module.ts:
```diff
@@ -9,6 +9,9 @@ import { MatChipsModule } from '@angular/material/chips';
 import { MatButtonModule } from '@angular/material/button';
+import { MatCardModule } from '@angular/material/card';
+
+import { DragDropModule } from '@angular/cdk/drag-drop';
```
Remove newline before.

d3lm reviewed (changes requested):
Forgot to mention that I'd like to move the drag & drop demo above the mobile demo.

Also, I found some more cosmetic issues. I know, that sucks but I cannot really automatically enforce newlines in scss files.

Can you again explain why range selection and CMD click doesn't work anymore with the drag and drop demo? Is that because Angular Material intercepts the click?

Commit:
refactor(app): moved drag-and-drop above mobile and hides
style(app): linting some scss

bvkimball commented:
No worries, updated.  I also added the `*ngIf="isDesktop"` to hide the `drag-and-drop` demo on mobile.

 Yeah the click to select functionality doesn't work because of the `DragDropModule` the `cdkDrag` directive captures all mouse events.   There are PRs, bugs, and feature requests for it as seen here: https://github.com/angular/components/pull/19674

It would be awesome if the `DragDropModule` didn't do this because then the 2 libraries pair together very well.

bvkimball commented:
I guess a solution could be to pass mouse-events from the selectItem as a "CustomEvent" or pass through to the parent via the reference we just added via DI.

d3lm commented:
Yea ok, that makes sense. Then I guess we can ignore this for now. It's fine because it's mostly related to the CDK then.

Tests are failing now. I think you have to update your tests to account for the desktop / mobile change.

Commit:
fix(app): fixed cypress test, error with test

bvkimball commented:
Yeah, Cypress error was because the "Item" animates while you drag to reorder, Because it isn't at the bottom of the screen the `scrollIntoView` causes some issue with the "drop location".  I changed the test to drag items "down" instead of up.  Tested manually too, the issue is just nuances with faking the events for testing.

d3lm commented:
So when I try the drag and drop I have this weird behavior when you move an item in the same list, then something's off with the placeholder. Would you mind checking this? Seems to be wrong behavior. Any idea why that is?

![CleanShot 2020-11-12 at 22 26 15](https://user-images.githubusercontent.com/12571019/98998548-9ac51100-2536-11eb-87ac-c9e38b785150.gif)

bvkimball commented:
ugh... yeah, i see it, it is caused because i hide `app-task` which has the `dtsSelectItem` but the `cdkDrag` container is is still in the dom...  I can fix this sometime today.

d3lm commented:
Awesome! Thanks. Looking forward to that fix. Then we are good to go I think.

Commit:
fix(app): visual issues with hidden placeholders

bvkimball commented:
Ok this fix is not great... Its kindof a hack, but since the `cdkDrag` really only manipulates the dom based on the element you are dragging and just hide the "other" selected items.  then those dom elements are still included in the re-ordering logic, i tried a bunch of things like make them absolute positioned, set height to 0, etc. the only thing that worked was to remove them from the dom and add them back after.  Once again this is an issue with the CDKDragModule and more of an issue of trying to do something it was not quite yet built for.  I think it is still great to show this still as an example of something a lot of people will want to do with this library.

Commit:
style(app): trying to force tests to re-run

promoter0100 commented:
I just come here when trying to use Drag & Drop with this lib, and I wonder if this function works?
